network/hybridNetwork.go

```diff
@@ -0,0 +1,222 @@
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// HybridP2PNetwork runs both P2PNetwork and WebsocketNetwork to implement the GossipNode interface
+type HybridP2PNetwork struct {
+	p2pNetwork *P2PNetwork
+	wsNetwork  *WebsocketNetwork
+	genesisID  string
+
+	useP2PAddress bool
+}
+
+// NewHybridP2PNetwork constructs a GossipNode that combines P2PNetwork and WebsocketNetwork
+func NewHybridP2PNetwork(log logging.Logger, cfg config.Local, datadir string, phonebookAddresses []string, genesisID string, networkID protocol.NetworkID, nodeInfo NodeInfo) (*HybridP2PNetwork, error) {
+	if !cfg.EnableP2PHybridMode {
+		return nil, fmt.Errorf("hybrid mode is not enabled")
+	}
+	wsnet, err := NewWebsocketNetwork(log, cfg, phonebookAddresses, genesisID, networkID, nodeInfo)
+	if err != nil {
+		return nil, err
+	}
+	// supply alternate NetAddress for P2P network
+	p2pcfg := cfg
+	p2pcfg.NetAddress = cfg.P2PListenAddress
+	p2pnet, err := NewP2PNetwork(log, p2pcfg, datadir, phonebookAddresses, genesisID, networkID)
+	if err != nil {
+		return nil, err
+	}
+	return &HybridP2PNetwork{
+		p2pNetwork: p2pnet,
+		wsNetwork:  wsnet,
+		genesisID:  genesisID,
+	}, nil
+}
+
+// Address implements GossipNode
+func (n *HybridP2PNetwork) Address() (string, bool) {
+	// TODO map from configuration? used for REST API, goal status, algod.net, etc
+	if n.useP2PAddress {
+		return n.p2pNetwork.Address()
+	}
+	return n.wsNetwork.Address()
+}
+
+type multiError struct{ p2pErr, wsErr error }
+
+func (e *multiError) Error() string   { return fmt.Sprintf("p2pErr: %s, wsErr: %s", e.p2pErr, e.wsErr) }
+func (e *multiError) Unwrap() []error { return []error{e.p2pErr, e.wsErr} }
+
+func (n *HybridP2PNetwork) runParallel(fn func(net GossipNode) error) error {
+	var wg sync.WaitGroup
+	var p2pErr, wsErr error
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		p2pErr = fn(n.p2pNetwork)
+	}()
+	go func() {
+		defer wg.Done()
+		wsErr = fn(n.wsNetwork)
+	}()
+	wg.Wait()
+
+	if p2pErr != nil && wsErr != nil {
+		return &multiError{p2pErr, wsErr}
+	}
+	if p2pErr != nil {
+		return p2pErr
+	}
+	if wsErr != nil {
+		return wsErr
+	}
+	return nil
+}
+
+// Broadcast implements GossipNode
+func (n *HybridP2PNetwork) Broadcast(ctx context.Context, tag protocol.Tag, data []byte, wait bool, except Peer) error {
+	return n.runParallel(func(net GossipNode) error {
```

Comment:
shouldn't we limit type of traffic broadcasted with p2p in hybrid mode? I'm afraid of xN bandwidth increase

Reply:
I agree about a constant number of peers, but I think we'd still need logic to prevent peers from "doubling up" and connecting to each other over both networks. If that logic's in this PR I haven't identified it yet